File: modules/expo-pose-detection/android/src/main/java/expo/modules/posedetection/ExpoPoseDetectionModule.kt
package expo.modules.posedetection

import android.content.Context
import android.util.Log
import androidx.core.os.bundleOf
import com.google.mediapipe.framework.image.MPImage
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.core.OutputHandler
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.mrousavy.camera.frameprocessors.FrameProcessorPluginRegistry
import expo.modules.posedetection.poselandmarksframeprocessor.PoseLandmarksFrameProcessorPlugin

class ExpoPoseDetectionModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw IllegalStateException("React Context not available")

  companion object {
    init {
      FrameProcessorPluginRegistry.addFrameProcessorPlugin("poseLandmarks") { proxy, options ->
        PoseLandmarksFrameProcessorPlugin(proxy, options)
      }
    }
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoPoseDetection")

    // Events
    Events(
      "onPoseLandmarksDetected",
      "onPoseLandmarksStatus",
      "onPoseLandmarksError"
    )

    // ADD THIS: Lifecycle callbacks for event observation
    OnStartObserving("onPoseLandmarksDetected") {
      Log.d("ExpoPoseDetection", "Started observing pose landmarks")
      // Ensure model is initialized when JavaScript starts listening
      if (PoseLandmarkerHolder.poseLandmarker == null) {
        initModel()
      }
    }

    OnStopObserving("onPoseLandmarksDetected") {
      Log.d("ExpoPoseDetection", "Stopped observing pose landmarks")
    }

    // CHANGE THIS: Make initModel synchronous and call it on module creation
    OnCreate {
      Log.d("ExpoPoseDetection", "Module created, initializing model")
      initModel()
    }

    Function("hello") {
      "Hello react native! 👋"
    }
  }

  private fun initModel() {
    // REMOVE THIS CHECK - allow re-initialization
    // This was preventing the model from reinitializing on camera switch
    if (PoseLandmarkerHolder.poseLandmarker != null) {
      Log.d("ExpoPoseDetection", "Model already initialized")
      sendEvent("onPoseLandmarksStatus", bundleOf("status" to "Model already initialized"))
      return
    }

    val resultListener = OutputHandler.ResultListener { result: PoseLandmarkerResult, inputImage: MPImage ->
      Log.d("PoseLandmarksFrameProcessor", "Detected ${result.landmarks().size} poses")

      val landmarksArray = result.landmarks().map { poseLandmarks ->
        poseLandmarks.mapIndexed { index, landmark ->
          mapOf(
            "keypoint" to index,
            "x" to landmark.x().toDouble(),
            "y" to landmark.y().toDouble(),
            "z" to landmark.z().toDouble(),
            "visibility" to landmark.visibility().orElse(0f).toDouble(),
            "presence" to landmark.presence().orElse(0f).toDouble()
          )
        }
      }

      sendEvent("onPoseLandmarksDetected", bundleOf("landmarks" to landmarksArray))
    }

    try {
      val baseOptions = BaseOptions.builder()
        .setModelAssetPath("pose_landmarker_lite.task")
        .build()

      val poseLandmarkerOptions = PoseLandmarker.PoseLandmarkerOptions.builder()
        .setBaseOptions(baseOptions)
        .setNumPoses(1)
        .setRunningMode(RunningMode.LIVE_STREAM)
        .setMinTrackingConfidence(0.8f)
        .setMinPoseDetectionConfidence(0.8f)
        .setMinPosePresenceConfidence(0.8f)
        .setResultListener(resultListener)
        .build()

      PoseLandmarkerHolder.poseLandmarker =
        PoseLandmarker.createFromOptions(context, poseLandmarkerOptions)

      Log.d("ExpoPoseDetection", "Model initialized successfully")
      sendEvent("onPoseLandmarksStatus", bundleOf("status" to "Model initialized successfully"))
    } catch (e: Exception) {
      Log.e("PoseLandmarksFrameProcessor", "Error initializing PoseLandmarker", e)
      sendEvent("onPoseLandmarksError", bundleOf("error" to (e.message ?: "Unknown error")))
    }
  }
}

File: modules/expo-pose-detection/ios/ExpoPoseDetectionModule.swift
import ExpoModulesCore
import MediaPipeTasksVision
import VisionCamera

// MARK: - Delegate Handler
class PoseLandmarkerResultProcessor: NSObject, PoseLandmarkerLiveStreamDelegate {
  
  weak var module: ExpoPoseDetectionModule?
  
  init(module: ExpoPoseDetectionModule) {
    self.module = module
    super.init()
  }
  
  func poseLandmarker(
    _ poseLandmarker: PoseLandmarker,
    didFinishDetection result: PoseLandmarkerResult?,
    timestampInMilliseconds: Int,
    error: Error?
  ) {
    
    if let error = error {
      print("Pose detection error: \(error.localizedDescription)")
      module?.sendEvent("onPoseLandmarksError", ["error": error.localizedDescription])
      return
    }
    
    guard let result = result else {
      print("No pose detection result received.")
      return
    }
    
    print("Detected \(result.landmarks.count) poses")
    
    let landmarksArray = result.landmarks.map { poseLandmarks in
      poseLandmarks.enumerated().map { (index, landmark) in
        [
          "keypoint": index,
          "x": Double(landmark.x),
          "y": Double(landmark.y),
          "z": Double(landmark.z),
          "visibility": Double(landmark.visibility?.floatValue ?? 0.0),
          "presence": Double(landmark.presence?.floatValue ?? 0.0)
        ] as [String: Any]
      }
    }
    
    module?.sendEvent("onPoseLandmarksDetected", ["landmarks": landmarksArray])
  }
}

// MARK: - Main Module
public class ExpoPoseDetectionModule: Module {
  
  private var resultProcessor: PoseLandmarkerResultProcessor?
  
  public func definition() -> ModuleDefinition {
    Name("ExpoPoseDetection")
    
    Events(
      "onPoseLandmarksDetected",
      "onPoseLandmarksStatus",
      "onPoseLandmarksError"
    )
    
    OnStartObserving("onPoseLandmarksDetected") {
      print("Started observing pose landmarks")
      if PoseLandmarkerHolder.shared.poseLandmarker == nil {
        self.initModel()
      }
    }
    
    OnStopObserving("onPoseLandmarksDetected") {
      print("Stopped observing pose landmarks")
    }
    
    OnCreate {
      print("Module created, initializing model")
      self.initModel()
        
        // Manually register frame processor
      FrameProcessorPluginRegistry.addFrameProcessorPlugin("poseLandmarks", withInitializer: { proxy, options in
            return PoseLandmarksFrameProcessorPlugin(proxy: proxy, options: options)
          })
    }
    
    OnDestroy {
      PoseLandmarkerHolder.shared.clear()
      self.resultProcessor = nil
      print("PoseLandmarker destroyed")
    }
    
    Function("hello") {
      return "Hello react native! 👋"
    }
  }
  
  private func initModel() {
    if PoseLandmarkerHolder.shared.poseLandmarker != nil {
      print("Model already initialized")
      sendEvent("onPoseLandmarksStatus", ["status": "Model already initialized"])
      return
    }
    
    do {
      resultProcessor = PoseLandmarkerResultProcessor(module: self)
      
      guard let modelPath = Bundle.main.path(forResource: "pose_landmarker_lite", ofType: "task") else {
        sendEvent("onPoseLandmarksError", ["error": "Model file not found"])
        print("Error: Model file not found in bundle")
        return
      }
      
      let options = PoseLandmarkerOptions()
      options.baseOptions.modelAssetPath = modelPath
      options.runningMode = .liveStream
      options.numPoses = 1
      options.minPoseDetectionConfidence = 0.8
      options.minTrackingConfidence = 0.8
      options.minPosePresenceConfidence = 0.8
      options.poseLandmarkerLiveStreamDelegate = resultProcessor
      
      try PoseLandmarkerHolder.shared.initializePoseLandmarker(with: options)
      
      sendEvent("onPoseLandmarksStatus", ["status": "Model initialized successfully"])
      print("Model initialized successfully")
    } catch {
      sendEvent("onPoseLandmarksError", ["error": error.localizedDescription])
      print("Error initializing PoseLandmarker: \(error.localizedDescription)")
    }
  }
}
